server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -469,13 +470,17 @@ long phase2(final long startingSeqNo, long requiredSeqNoRangeStart, long endingS
             "required [" + requiredSeqNoRangeStart + ":" + endingSeqNo + "]");
 
         // send all the snapshot's translog operations to the target
-        final SendSnapshotResult result = sendSnapshot(startingSeqNo, requiredSeqNoRangeStart, endingSeqNo, snapshot);
-
-        stopWatch.stop();
-        logger.trace("recovery [phase2]: took [{}]", stopWatch.totalTime());
-        response.phase2Time = stopWatch.totalTime().millis();
-        response.phase2Operations = result.totalOperations;
-        return result.targetLocalCheckpoint;
+        try {
+            final SendSnapshotResult result = sendSnapshot(startingSeqNo, requiredSeqNoRangeStart, endingSeqNo, snapshot);
+            stopWatch.stop();
+            logger.trace("recovery [phase2]: took [{}]", stopWatch.totalTime());
+            response.phase2Time = stopWatch.totalTime().millis();
+            response.phase2Operations = result.totalOperations;
+            return result.targetLocalCheckpoint;
+        } catch (TranslogCorruptedException ex) {
```

Comment:
I think this should somehow be folded into the snapshot reading. This means the engine will be failed regardless on who uses the snapshot.